### DISCUSSION TITLE
OpenGL Functions not Declared

### DISCUSSION THREAD

**Discussion by Ultim8Xeno** [ocornut/imgui#4896] [New Users Build/Link/Run issues ONLY!]
so I tried to add imgui to my project and I'm getting a bunch of errors relating to undeclared/undefined functions
Here are the errors : https://paste.myst.rs/qisz3no7
I'm calling these imgui commands to begin and end : https://paste.myst.rs/kz9yzf0u
The error is that in "imgui_impl_opengl3_loader.h", a few opengl functions aren't defined.
Those functions are defined in this : https://paste.myst.rs/1fuhleqo
Which is locked behind the ifndef guard, meaning GL_VERSION_1_1 is defined somewhere
I tried putting a "#undef GL_VERSION_1_1" before the guard (and a #define right after to make sure I don't break anything). That did fix the errors but the imgui windows appear behind all the other things I render. That might just be a problem with something else but I'm not sure (I'm using GL_DEPTH_TEST).
Where and why is GL_VERSION_1_1 defined and why does that effect the functions in that block? Thanks for the help

**Comment by PathogenDavid** [comment]
Can you share the contents of `ImGuiBase.cpp`? It looks like you're including `imgui_impl_opengl_loader.h` directly; which you shouldn't be doing since it's internal to the OpenGL backend and not meant for general consumption.

**Reply by Ultim8Xeno** [reply]
ImGuiBase is the second link. I'm not including imgui_impl_opengl_loader.h directly, I am however including the impl .cpp and .h files in ImGuiBase.cpp. Also, I don't know if this does anything but I'm using premake and ImGui is it's own visual studio project which I'm linking as a static lib

**Reply by PathogenDavid** [reply]
> ImGuiBase is the second link.

Oops, I somehow missed that.

> I am however including the impl .cpp and .h files in ImGuiBase.cpp.

Is there a reason you're doing this instead of just building it in your static library?

(Also FYI, `IMGUI_IMPL_OPENGL_LOADER_GLAD` no longer does anything since Dear ImGui has its own loader.)

**Reply by Ultim8Xeno** [reply]
I tried doing that but it gives me unresolved external errors for functions in the cpp files
Here's my premake file if that helps : https://paste.myst.rs/ifsls4wi

**Reply by PathogenDavid** [reply]
```lua
"backends/imgui_impl_opengl3.cpp>"
```

You have a typo here. If that doesn't resolve the errors, can you share what they are?

**Reply by Ultim8Xeno** [reply]
oh didn't realize since I copy pasted from the include
Anyways, it works :D Thank you so much!
One more thing, the window does still appear behind everything else I render, so it doesn't have anything to do with my previous problem.
I think I might be able to figure that one out on my own though so that's all the help I need
Again, tysm!

**Reply by PathogenDavid** [reply]
No problem!

> One more thing, the window does still appear behind everything else I render

You probably just need to make sure `ImGui_ImplOpenGL3_RenderDrawData` is called after rendering the other things. The OpenGL backend doesn't use depth since it's not super meaningful in a 2D context.
